docs/examples/frameworks/pytorch/pytorch-lightning.ipynb

```diff
@@ -0,0 +1,488 @@
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "1"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model = LitMNIST()\n",
+    "trainer = Trainer(gpus=1, distributed_backend=\"ddp\", max_epochs=5)\n",
+    "trainer.fit(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let us define a DALI pipeline which would load the data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import nvidia.dali as dali\n",
+    "from nvidia.dali.pipeline import Pipeline\n",
+    "import nvidia.dali.ops as ops\n",
+    "import nvidia.dali.types as types\n",
+    "from nvidia.dali.plugin.pytorch import DALIClassificationIterator\n",
+    "\n",
+    "# Path to MNIST dataset\n",
+    "data_path = os.path.join(os.environ['DALI_EXTRA_PATH'], 'db/MNIST/training/')\n",
+    "\n",
+    "class MnistPipeline(Pipeline):\n",
+    "    def __init__(self, batch_size, device, device_id=0, shard_id=0, num_shards=1, num_threads=4, seed=0):\n",
+    "        super(MnistPipeline, self).__init__(\n",
+    "            batch_size, num_threads, device_id, seed)\n",
+    "        self.device = device\n",
+    "        self.reader = ops.Caffe2Reader(path=data_path, shard_id=shard_id, num_shards=num_shards, random_shuffle=True)\n",
+    "        self.decode = ops.ImageDecoder(\n",
+    "            device='mixed' if device == 'gpu' else 'cpu',\n",
+    "            output_type=types.GRAY)\n",
+    "        self.cmn = ops.CropMirrorNormalize(\n",
+    "            device=device,\n",
+    "            dtype=types.FLOAT,\n",
+    "            std=[0.3081 * 255],\n",
+    "            mean=[0.1307 * 255],\n",
+    "            output_layout=\"CHW\")\n",
+    "        self.to_int64 = ops.Cast(dtype=types.INT64, device=device)\n",
+    "\n",
+    "    def define_graph(self):\n",
+    "        inputs, labels = self.reader(name=\"Reader\")\n",
+    "        images = self.decode(inputs)\n",
+    "        images = self.cmn(images)\n",
+    "        if self.device == \"gpu\":\n",
+    "            labels = labels.gpu()\n",
+    "        # PyTorch expects labels as INT64\n",
+    "        labels = self.to_int64(labels)\n",
+    "\n",
+    "        return (images, labels)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Add DALI to the data preparation step in the training class and adjust the `process_batch` to accept the data returned by DALIClassificationIterator, which returns a list of dictionaries, where each list element corresponds to one pipeline wrapped by the DALIIterator, and entries in the dictionary corresponds to the relevant outputs. Check for details in the DALIGenericIterator documenation."
```

Comment:
Now we are ready to modify the training class to use the DALI pipeline we have just defined. Because we want to integrate with PyTorch, we wrap our pipeline with a PyTorch DALI iterator, that can replace the native data loader with some minor changes in the code. The DALI iterator returns a list dictionaries, where each element in the list corresponds to a pipeline instance, and the entries in the dictionary map to the outputs of the pipeline. For more information, check the documentation of DALIGenericIterator.

Reply:
Done